### PR TITLE
msg/async: add timeout for connections which are not ready

### PR DESCRIPTION
**Pull request by xiexingguo**
There could be various corner cases that may cause an async
connection stuck in the connecting stage (e.g., by manually
creating some loop back connections on the switches of our test cluster,
we can almost 100% reproduce http://tracker.ceph.com/issues/37499).
    
In 61b9432ef9a3847eceb96f8d5a854567c49bbf61 I try to employ the
existing keep_alive mechanism to get those stuck connections out of the
trap but it does not work if the corresponding connection
is not yet ready, since we always require the underlying connection to be
**ready** in order to send out a keep_alive message.
    
Fix by making a more general connecting timeout strategy.
If a connecting process can not be finished within a specific interval,
then we simply cut it off and retry.

Also renamed ms_tcp_read_timeout to ms_connection_idle_timeout,
since the old naming is super confusing.
    
Fixes: http://tracker.ceph.com/issues/37499
Fixes: http://tracker.ceph.com/issues/38493

- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

### CODE REVIEW

**Comment by xiexingguo** [comment]
Manually verified in our test lab, fixed http://tracker.ceph.com/issues/38493 and works well (for now).

File: src/common/options.cc [hunk 1]
```diff
@@ -955,9 +955,14 @@ std::vector<Option> get_global_options() {
     .set_default(512)
     .set_description("Size of queue of incoming connections for accept(2)"),
 
-    Option("ms_tcp_read_timeout", Option::TYPE_UINT, Option::LEVEL_ADVANCED)
+
+    Option("ms_connection_ready_timeout", Option::TYPE_UINT, Option::LEVEL_ADVANCED)
+    .set_default(600)
```
Comment (liewegas) [inline src/common/options.cc]: Ten minutes seems like a long time to be attempting to open a connection... it seems like we'd want to fault and retry/reconnect after maybe 10s of seconds at most?  Or am I misunderstanding the situations where this timeout comes up?

**Comment by tchaikov** [comment]
@xiexingguo could you rerun ceph_test_msgr to see if following failure is reproducible ?

http://pulpito.ceph.com/kchai-2019-04-07_01:07:57-rados-wip-kefu-testing-2019-04-05-1411-distro-basic-mira/3818231/

File: src/common/options.cc (hunk 1 above)
Comment (xiexingguo) [reply src/common/options.cc]: @liewegas 10s is fine with me:-)  10 minutes might be over-conservative...

**Comment by xiexingguo** [comment]
> could you rerun ceph_test_msgr to see if following failure is reproducible ?
> 
> http://pulpito.ceph.com/kchai-2019-04-07_01:07:57-rados-wip-kefu-testing-2019-04-05-1411-distro-basic-mira/3818231/

@tchaikov Sure. I'll check that failure.

**Comment by xiexingguo** [comment]
http://pulpito.ceph.com/xxg-2019-04-08_02:55:31-rados-wip-kefu-testing-2019-04-05-1411-distro-basic-smithi/

http://pulpito.ceph.com/xxg-2019-04-08_04:19:58-rados-wip-kefu-testing-2019-04-05-1411-distro-basic-smithi/

http://pulpito.ceph.com/xxg-2019-04-08_05:12:47-rados-wip-kefu-testing-2019-04-05-1411-distro-basic-smithi/

Try to reproduce that failure but with no luck. Hit http://tracker.ceph.com/issues/38577 again :-(

**Comment by tchaikov** [comment]
- http://pulpito.ceph.com/kchai-2019-04-07_01:07:57-rados-wip-kefu-testing-2019-04-05-1411-distro-basic-mira/
- http://pulpito.ceph.com/kchai-2019-04-08_08:52:22-rados-wip-kefu-testing-2019-04-05-1411-distro-basic-smithi/#

**Comment by tchaikov** [comment]
@liewegas good to merge?

**Comment by liewegas** [comment]
That mira run has a lot of noise.. are there normally that many transient failures on mira?

**Comment by tchaikov** [comment]
@liewegas yes, that's normal.

**Pull request merged**